Add m_checkbans

genius3000 opened the pull request:
Provides three commands: /checkbans, /testban, and /whyban
`checkbans`: Check a channel's ban/exception list against its current users
`testban`: Test a ban mask against the current users in a channel
`whyban`: See what bans/exceptions match yourself or another user to a channel

I'm thinking the numerics currently used should be changed, as some clients probably won't like seeing the Banlist and End of Banlist numerics used here. Any suggestions?

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		/* Check for channel mode 'e', ban exceptions */
+		ListModeBase* ex = (ListModeBase*)ServerInstance->Modes->FindMode('e', MODETYPE_CHANNEL);
```
You should check the mode name here before casting it as 'e' might be bound to another mode.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+
+		/* Check for channel mode 'e', ban exceptions */
+		ListModeBase* ex = (ListModeBase*)ServerInstance->Modes->FindMode('e', MODETYPE_CHANNEL);
+		modelist* exceptions = NULL;
```
This could just be something like

```cpp
modelist* exceptions = ex ? ex->extItem->get(c) : NULL;
```

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+			{
+				if (c->CheckBan(u->first, ban->data))
+				{
+					user->WriteServ("367 %s %s Ban: %s matches %s (set by: %s on %s)", user->nick.c_str(),
```
1. This numeric collides with other numerics already used by InspIRCd. Pick something else that [is not already in use](https://defs.ircdocs.horse/defs/numerics.html).

2. This is not a well formed IRC message. You should prefix the final text parameter with a `:` to make it into a <trailing> parameter.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+				}
+			}
+
+			if (!exceptions)
```
If `exceptions` is not used until here then you should bring it down here instead of leaving it up top.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+			{
+				if (c->CheckBan(u->first, exc->mask))
+				{
+					user->WriteServ("367 %s %s Exception: %s matches %s (set by: %s on %s)", user->nick.c_str(),
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+			}
+		}
+
+		user->WriteServ("368 %s %s :End of check bans list", user->nick.c_str(), c->name.c_str());
```
The same as above re: numeric usage applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		{
+			if (c->CheckBan(u->first, parameters[1]))
+			{
+				user->WriteServ("367 %s %s Ban: %s matches %s", user->nick.c_str(), c->name.c_str(),
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		if (matched > 0)
+		{
+			float percent = ((float)matched / (float)users->size()) * 100;
+			user->WriteServ("367 %s %s Ban: %s matched %d of %lu users (%.2f%%).", user->nick.c_str(), c->name.c_str(),
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+						parameters[1].c_str(), matched, users->size(), percent);
+		}
+
+		user->WriteServ("368 %s %s :End of test ban list", user->nick.c_str(), c->name.c_str());
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		 * allow checking others if you have access to the channel ban list.
+		 */
+		User* u;
+		if (parameters.size() == 1)
```
```cpp
User* u == parameters.size() == 1 ? user : NULL;
```

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		{
+			if (c->CheckBan(u, ban->data))
+			{
+				user->WriteServ("367 %s %s Ban: %s (set by: %s on %s)", user->nick.c_str(), c->name.c_str(),
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		/* Now try to get the exceptions list and check it */
+		ListModeBase* ex = (ListModeBase*)ServerInstance->Modes->FindMode('e', MODETYPE_CHANNEL);
```
The same as above re: checking the mode name.

SaberUK commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,274 @@
+			{
+				if (c->CheckBan(u, exc->mask))
+				{
+					user->WriteServ("367 %s %s Exception: %s (set by: %s on %s)", user->nick.c_str(),
```
The same as above re: numeric usage and message validity applies here too.

SaberUK commented:
You might want to consider checking if you can deduplicate some of the code in this as theres quite a bit which seems to be duplicated.

bigfoot547 commented:
Could this module be used to decloak people?
If so, could there be some flood penalty applied?

genius3000 commented:
> You might want to consider checking if you can deduplicate some of the code in this as theres quite a bit which seems to be duplicated. --@SaberUK 

I pondered that a bit too before, I'll see what I can do though. Also thanks for the reviews, I'll find unused numerics and name them; as well as clean up the messages.

> Could this module be used to decloak people?
If so, could there be some flood penalty applied? --@bigfoot547 

Yeah it could be used to decloak, limiting to halfop/op helps some but I'll add a flood penalty as well and look into any other options to limit/prevent abusing this. Thanks for the catch.

genius3000 commented:
Alright I rebased and added a commit that should take care of all the comments/concerns.  
Managed to deduplicate the main listing code, exception mode check (also fixed the casting before knowing what it is issue), and the 'need chanops' message.  
Found unused numeric numbers to use and named them, fixed the messages, and added a penalty of 6 to both /checkbans and /testban, with a penalty of 10 to /whyban <nick>
I'll squash the commits once approved.

linuxdaemon commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,263 @@
+		User* u = parameters.size() == 1 ? user : NULL;
+		if (!u)
+		{
+			this->Penalty = 10;
```
I believe what you meant here is `user->CommandFloodPenalty += 10000;` to set a user's penalty. This will currently modify the penalty for the command if anyone attempts to check another user.

genius3000 replied:
If I recall correctly, that's exactly my intention. The command handler will then assign the penalty.

linuxdaemon replied:
Yes but this currently sets the penalty for the command no matter what, after that line, any time the command is ran a penalty of 10 will be applied.

genius3000 replied:
Ah yeah I see what you're saying now. Thanks.

linuxdaemon commented on 2.0/m_checkbans.cpp:
```diff
@@ -0,0 +1,263 @@
+		// Check for matching bans and exceptions (if available)
+		CheckLists(user, c, u, ex);
+
+		user->WriteNumeric(RPL_ENDLIST, "%s %s :End of why ban list", user->nick.c_str(), c->name.c_str());
```
Might be better to do something like `nick channel checked_nick :End of /WHYBAN list` to mimic how commands like `WHO` respond.
